src/OpenTelemetry/Trace/ActivityProcessor.cs

```diff
@@ -114,8 +114,12 @@ protected virtual void Dispose(bool disposing)
             this.disposed = true;
         }
 
-        protected abstract void OnStartInternal(Activity activity);
+        protected virtual void OnStartInternal(Activity activity)
```

Comment:
@alanwest The "Internal" in the name is kind of non-standard. I think we were talking about this before but I lost track. Could we not do?
```csharp
public void Start(Activity activity)
protected virtual void OnStart(Activity activity)
```

Reply:
I haven't looked through the changes yet, imagine we have a chain of processors, checking `SuppressInstrumentation` on every single processor seems to be a big perf hit.

Reply:
> @alanwest The "Internal" in the name is kind of non-standard. I think we were talking about this before but I lost track. Could we not do?

Agreed. I actually have this change staged up, just haven't pushed yet. Mostly just picked things on the PR today and resolved the various conflicts.

> I haven't looked through the changes yet, imagine we have a chain of processors, checking SuppressInstrumentation on every single processor seems to be a big perf hit.

Yes, I haven't spent enough time analyzing this yet. Earlier there was an idea from @CodeBlanch to see if this fits better in the sampler. Will consider this soon.

Reply:
Given `SuppressInstrumentation` is used for folks to express whether they want the telemetry to be suppressed or not, how about the following approach:
* If the code (app/lib) is using new `ActivitySource.StartActivity`, it will be responsible to check the `SuppressInstrumentation` flag and not creating the activity if the flag is set.
  * In this way the sampler/processor/exporter won't receive the activity at all since it is not even created.
  * If the code is not performing the check, the code is faulty and we are noting doing another round of check in the processor - due to perf reason.
* If the code is using legacy Activity, the adapter will be responsible to drop the activity before sending it to the sampler/processor.
  * In this way we gain perf since the adapter is only taking legacy activities.

Reply:
> If the code (app/lib) is using new ActivitySource.StartActivity, it will be responsible to checking the SuppressInstrumentation flag and not creating the activity if the flag is set.

This would require taking an OTel dependency. We want to avoid this need, no?

> If the code is using legacy Activity, the adapter will be responsible to drop the activity before sending it to the sampler/processor.

Yes, this makes sense for the legacy route.

Reply:
I would worry a bit relying on others to do it perfectly. I'm envisioning a bunch of difficult-to-triage issues being opened when applications go into infinite loops. We could move the check into the ActivityListener? If SuppressInstrumentation is triggered, we return ActivityDataRequest.None from GetRequestedDataUsingContext (or whatever it ends up being called). If we do that, Activity is never created.

Reply:
> We could move the check into the ActivityListener?

If perf looks good, putting the logic in the ActivityListener will work.

> This would require taking an OTel dependency. We want to avoid this need for library authors, no?

Definitely yes. I think most library authors don't need to care about this (infinite loops). Only library authors who build libraries that can be used for exporting task will have to worry about it. Probably on for HTTP / TCP / UDP / Unix domain socket.

Reply:
More data points:
* In OpenTelemetry Python it seems to be fine to have only HTTP instrumentation doing the loop detection using SuppressInstrumentation, check [this](https://github.com/open-telemetry/opentelemetry-python/search?q=suppress&unscoped_q=suppress).
* If only 5% of the scenario would use it, making it too generic and ask everyone to pay the perf tax could be a hard-sell (if the perf tax is low, for example 0.1%, I guess it'll be fine).

Reply:
I'll try on this on for size tomorrow:
* Check in `ActivityListener` for the new `ActivitySource` way
* And check in the adapter for the legacy

Then try to get some perf numbers. If we're not happy, then I'm definitely ok entertaining scoping this only to certain instrumentation `HttpClient`/`HttpWebRequest`.

Reply:
@alanwest FYI - I've shared a mini benchmark project which I used for refactor the provider/processor/pipeline https://github.com/open-telemetry/opentelemetry-dotnet/pull/1039.